File: composeApp/src/androidMain/kotlin/io/chuckstein/doable/tracker/TrackerViewModel.kt
package io.chuckstein.doable.tracker

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.chuckstein.doable.tracker.TrackerEvent.InitializeTracker

class TrackerViewModel(private val stateEngine: TrackerStateEngine) : ViewModel() {

    init {
        processEvent(InitializeTracker)
    }

    val uiState = stateEngine.collectUiState(viewModelScope)

    fun processEvent(event: TrackerEvent) {
        stateEngine.processEvent(event, viewModelScope)
    }
}

File: iosApp/iosApp/tracker/TrackerViewModel.swift
import Foundation
import Combine
import ComposeApp

@MainActor
class TrackerViewModel: ObservableObject {

    let stateEngine: TrackerStateEngine = KoinHelper().stateEngine
    let viewModelScope = ViewModelScope()
    
    @Published var uiState: TrackerUiState

    init() {
        self.uiState = stateEngine.currentUiState()
        
        processEvent(TrackerEventInitializeTracker())

        // this and the @Published uiState is not actually used anymore
        stateEngine.onUiStateChange(scope: viewModelScope) { [weak self] newState in
            self?.uiState = newState
        }
    }

    func processEvent(_ event: TrackerEvent) {
        stateEngine.processEvent(event: event, scope: viewModelScope)
    }

    deinit {
        viewModelScope.cancel()
    }
}
